pkg/client/events/query_client.go

```diff
@@ -107,70 +119,115 @@ func (eqc *eventsQueryClient) EventsBytes(
 	defer eqc.eventsBytesAndConnsMu.Unlock()
 
 	// Check if an event subscription already exists for the given query.
-	if eventsBzConn := eqc.eventsBytesAndConns[query]; eventsBzConn != nil {
+	if eventsBzConn := eqc.eventsBytesInfo[query]; eventsBzConn != nil {
 		// If found it is returned.
 		return eventsBzConn.eventsBytes, nil
 	}
 
+	// Check and close any existing connection for the given query.
+	// This ensures that we do not have multiple subscriptions for the same query.
+	// This is to prevent duplicate event notifications.
+	if oldConn, ok := eqc.eventsBytesInfo[query]; ok {
+		// If an old connection exists for the given query, close it.
+		eqc.logger.Warn().Msgf(
+			"♻️ Replacing existing event subscription for query %s and id %d with a new one",
+			query,
+			oldConn.connId,
+		)
+
+		// cancel the old context to stop the unsubscribe goroutine
+		oldConn.cancelCtx()
+	}
+
 	// Otherwise, create a new event subscription for the given query.
 	eventsBzConn, err := eqc.newEventsBytesAndConn(ctx, query)
 	if err != nil {
 		return nil, err
 	}
 
 	// Insert the new eventsBytes into the eventsBytesAndConns map.
-	eqc.eventsBytesAndConns[query] = eventsBzConn
+	eqc.eventsBytesInfo[query] = eventsBzConn
 
 	// Unsubscribe from the eventsBytes when the context is done.
-	go eqc.goUnsubscribeOnDone(ctx, query)
+	go eqc.goUnsubscribeOnDone(eventsBzConn)
 
 	// Return the new eventsBytes observable for the given query.
 	return eventsBzConn.eventsBytes, nil
 }
 
 // Close unsubscribes all observers from all event subscription observables.
 func (eqc *eventsQueryClient) Close() {
-	eqc.close()
+	for _, eventsBzConn := range eqc.eventsBytesInfo {
+		eqc.close(eventsBzConn)
+	}
 }
 
 // close unsubscribes all observers from all event subscription observables.
-func (eqc *eventsQueryClient) close() {
+func (eqc *eventsQueryClient) close(evtConn *eventsBytesInfo) {
 	eqc.eventsBytesAndConnsMu.Lock()
 	defer eqc.eventsBytesAndConnsMu.Unlock()
 
-	for query, eventsBzConn := range eqc.eventsBytesAndConns {
-		// Unsubscribe all observers of the eventsBzConn observable and close the
-		// connection for the given query.
-		eventsBzConn.Close()
-		// remove isClosed eventsBytesAndConns
-		delete(eqc.eventsBytesAndConns, query)
+	if _, ok := eqc.eventsBytesInfo[evtConn.query]; ok {
+		// mark the connection as closed
```

Comment:
Why do we need to do this rather than just close it?

There seems to be important magic happening here, but I don't get it.